pyproject.toml

```diff
@@ -32,7 +35,13 @@ dev = [
     "pre-commit>=3.8.0",
     "pytest>=8.3.2",
     "pooch>=1.8.2",
+    "ipdb>=0.13.13",
+    "gpustat>=1.1.1",
+    "zarrdump>=0.4.1",
+    "dask[distributed]>=2024.7.1",
+    "bokeh!=3.0.*,>=2.4.2",
```

Comment:
Except for dask, are these new dev dependencies used? I could not find them imported or referenced anywhere else.

Reply:
I use `ipdb` for interactive debugging while I am developing the code, `gpustat` while I run training (since it gives me gpu utilisation) and `zarrdump` is a useful command line tool for printing the contents of zarr files. So they aren't used in `neural-lam` directly no, but I use them while developing the code. We can remove them from the dependencies if you like, but I they are just in my standard toolchain for development which is why I thought they fit in the `dev` group. Otherwise I have to install them with `pip` manually each time I work on the codebase.

`dask[distributed]` and `bokeh` are a bit more tricky. They are needed if one wants to preprocess the training dataset with `mllam-data-prep` using `dask.distributed` and `bokeh` provides the visualisation (https://github.com/mllam/mllam-data-prep?tab=readme-ov-file#creating-large-datasets-with-daskdistributed). I don't include them by default in `mllam-data-prep` because they aren't necessary for making a dataset, but can make it easier/nicer. There I have added a optional dependecy group for those two packages https://github.com/mllam/mllam-data-prep/blob/main/pyproject.toml#L26. Maybe I should add an optional dependencies group to `neural-lam` too, something like `neural-lam[mdp-dask-distributed]`. I'm not sure what is best...

Reply:
To me it seems reasonable that tools that we personally use for development we have to install manually using pip. No point in "forcing" those tools on others working on the code.

Yes I can see how dask[distributed] and bokeh are tricky. But if I understand correctly what these are needed for is not related to development, so it seems to fit poorly under dev. Could an option just be to refer people to installing these from mllam-data-prep? So if you want these features you would then just install mllam-data-prep[dask-distributed]? Otherwise I think the neural-lam[mdp-dask-distributed] option would also work, but risks having to maintain this list twice. Or maybe one could just have mllam-data-prep[dask-distributed] as the only dependency in neural-lam[mdp-dask-distributed].

Reply:
I suggest the following solution for now, which I have tested with `pdm install --group dev` and it works. I tend to agree that we should not enforce dev-tools on the user.
```
dependencies = [
    "numpy>=1.24.2",
    "wandb>=0.13.10",
    "scipy>=1.10.0",
    "pytorch-lightning>=2.0.3",
    "shapely>=2.0.1",
    "networkx>=3.0",
    "Cartopy>=0.22.0",
    "pyproj>=3.4.1",
    "tueplots>=0.0.8",
    "matplotlib>=3.7.0",
    "plotly>=5.15.0",
    "torch>=2.3.0",
    "torch-geometric==2.3.1",
    "parse>=1.20.2",
    "dataclass-wizard>=0.22.3",
    "mllam-data-prep[dask-distributed] @ git+https://github.com/mllam/mllam-data-prep",
]
requires-python = ">=3.9"

[project.optional-dependencies]
dev = ["pre-commit>=3.8.0", "pytest>=8.3.2", "pooch>=1.8.2"]
```